tsdb: Block Head GC till pending readers are done reading

codesome opened the pull request:
Closes #5877 #8221 #9079

This PR tracks the in-flight readers on Head block and block Head truncation till those readers are done reading. And while the in-memory truncation is happening (and not the m-mapped stuff), getting querier on Head block is blocked.

I have copied the tests from #9077 and #9080 into this while modifying slightly to take care of blocking Compact()/Querier() calls.

~I have opened this as draft since the current approach is very naive without considering the time ranges of the queriers. I will be looking at improving it further to only block (and get blocked on) the queriers that would be affected with the truncation.~

~Only blocks memory truncation on overlapping queries and if truncation is going on or has started, only block queries that are overlapping with the truncation.~

The querier adjusts the time range based on the on-going truncation without taking a lock. The memory truncation waits for in-flight queries that overlap with the truncation to finish.

roidelapluie commented:
Couldn't we reuse the isolation tracking here?

codesome commented:
> Couldn't we reuse the isolation tracking here?

I think we can, we will need to additionally store and mint and maxt of the reads. Let me try that.

codesome commented:
I am now using the existing isolation tracking with mint and maxt in the isolation states. I have also added enhancements to only block memory truncation on overlapping queries _and_ if truncation is going on or started, only block queries that are overlapping with the truncation.

The PR is ready for review now. I will try adding more tests in case any part of it is untested yet.

codesome commented:
I have added more tests cases now with some small improvements and fixes. _Now_ it is finally ready for review.

roidelapluie commented on tsdb/head.go:
```diff
@@ -934,18 +941,89 @@ func (h *Head) Truncate(mint int64) (err error) {
+	h.memTruncateMtx.Unlock()
+}
+
+func (h *Head) WaitForPendingReadersInTimeRange(mint, maxt int64) {
```
This seems like an issue to me as it would block for up to 2 minutes, while we do not execute new queries. Could we think about an alternative?

codesome commented:
I have updated the PR with a better lock-free approach for this, and much simpler to understand. I have added detailed comments in the code explaining what is it trying to do. PR is ready for review again.

beorn7 commented on tsdb/isolation.go:
```diff
@@ -124,6 +127,21 @@ func (i *isolation) State() *isolationState {
 }
 
+// TraverseOpenReads iterates through the open reads and runs the given
+// function on those states. The given function SHOULD NOT mutate the isolationState.
```
Perhaps better MUST NOT? (o:

beorn7 commented on tsdb/db.go:
```diff
@@ -1516,8 +1516,29 @@ func (db *DB) Querier(_ context.Context, mint, maxt int64) (storage.Querier, err
+			return nil, errors.Wrapf(err, "open querier for head %s", rh)
+		}
+
+		shouldClose, getNew, newMint := db.head.IsQuerierValid(mint, maxt)
```
I might be missing a locking layer, but it looks to me like a truncation could start just after this line has returned "all good" (i.e. `false, false, 0`), and then we would run into the same problem again. (In other words, is still a race, just that we reduced the time window in which the race could be triggered.)

codesome replied:
Good question. So that is taken care by first getting the querier above before calling this (as getting querier will register itself in the isolation state linked list). So if we got `false, false, 0`, that means truncation had not started yet. So even if truncation starts after this, it will wait at `WaitForPendingReadersInTimeRange` since querier was registered before this point.

codesome replied:
I will add a comment in the code regarding this.

beorn7 replied:
OK, that might be the "locking layer" I was missing. I will have another look and try to wrap my head around it.

beorn7 reviewed (commented):
Hi @codesome , thanks for doing this. I have tried to fully understand this, but so far I have only gained a partial understanding. I have written down a few concerns and questions. Perhaps they are dumb, coming from my merely partial or even wrong understanding, but at least then I have the opportunity to improve my understanding.

codesome commented on tsdb/head.go:
```diff
@@ -987,6 +1004,76 @@ func (h *Head) truncateMemory(mint int64) (err error) {
 }
 
+func (h *Head) WaitForPendingReadersInTimeRange(mint, maxt int64) {
+	// TODO: should we have some limit on how long we wait in total? If so, how long?
```
Not forever. It will be blocked till all the queriers that overlap with this truncation are closed that were got _before_ truncation started. So it depends on how long the queries run. Should be few mins max before they are timed out?

beorn7 replied:
Yeah, so that directly depends on the query timeout, which is set via the `--query.timeout` flag, 2m by default. (It can be set on a per-query basis via the HTTP API, but it is still capped by the value set via the flag.) If the query timeout is really limiting the wait time here implicitly, I don't think we need to introduce another separate timeout here.

beorn7 replied:
Meaning: Perhaps remove the TODO and change it into an explanation that the query timeout limits the wait time here implicitly.

codesome commented:
@beorn7 those are all valid questions and I agree few things can get confusing. I have answered them in the comments.

beorn7 commented on tsdb/head.go:
```diff
@@ -987,6 +1004,72 @@ func (h *Head) truncateMemory(mint int64) (err error) {
+	}
+}
+
+// IsQuerierCollidingWithTruncation tells if the current querier should be closed and if
+// have to get a new querier. If should get a new querier, it also tells
+// what is the new mint of the Head to be considered.
+// This function helps in preventing race with truncation of in-memory data.
```
Suggestion to iron out a few language quirks:

"""
IsQuerierCollidingWithTruncation returns if the current querier needs to be closed and if a new querier has to be created. In the latter case, the method also returns the new mint to be used for creating the new range head and the new querier. This methods helps preventing races with the truncation of in-memory data.
"""

I also recommend to leave an empty comment line after this (`// `) because Godoc will otherwise render the `NOTE` below as part of the same paragraph, reducing its visibility.

beorn7 reviewed (approved):
I cannot say I'm 100% confident I have understood everything, but I guess it's good enough to give approval from my side. Just a few nits about doc comments.

codesome commented:
Thanks for the reviews! I made 1 additional fix to take care of exclusiveness of the truncation time in `WaitForPendingReadersInTimeRange` and added a test case for it. I will merge this on green.
